File: 백준/Gold/2624. 동전 바꿔주기/동전 바꿔주기.py
from sys import *
from collections import * 
from heapq import * 
from functools import *
from math import *
from itertools import *

#setrecursionlimit(10**9)
input = stdin.readline

t = int(input())
k = int(input())

coins = []

for _ in range(k):
    p,n = map(int,input().split())
    coins.append([p,n])
    

dp = [0] *(t+1)

dp[0] = 1
for cost,num in coins:
    for money in range(t,0,-1):
        for i in range(1,num+1):
            if money - cost * i >= 0:
                dp[money] += dp[money - cost *i]

print(dp[t])

File: 백준/Gold/2624. 동전 바꿔주기/동전 바꿔주기.swift
var t = Int(readLine()!)!
let n = Int(readLine()!)!

var coins: [(Int,Int)] = []



for _ in 0..<n {
    let ca = readLine()!.split{$0 == " "}.map{Int($0)!}
    
    coins.append((ca[0],ca[1]))
    
}

var dp = [Int](repeating: 0, count: t+1)

dp[0] = 1

for (coin,amount) in coins{
    
    for money in stride(from: t, to: 0, by: -1) {
        
        for i in 1...amount {
            
            if money - (coin*i) >= 0 {
                dp[money] += dp[money-coin*i]
            }
            
        }
        
    }

}

print(dp[t])

